// ol/render/webgl/texturereplay/defaultshader.js
// This file is automatically generated, do not edit
/* eslint openlayers-internal/no-missing-requires: 0 */
import _ol_ from '../../../index';
import _ol_webgl_Fragment_ from '../../../webgl/fragment';
import _ol_webgl_Vertex_ from '../../../webgl/vertex';
var _ol_render_webgl_texturereplay_defaultshader_ = {};

if (false) {

  /**
   * @constructor
   * @extends {ol.webgl.Fragment}
   * @struct
   */
  _ol_render_webgl_texturereplay_defaultshader_.Fragment = function() {
    _ol_webgl_Fragment_.call(this, _ol_render_webgl_texturereplay_defaultshader_.Fragment.SOURCE);
  };
  _ol_.inherits(_ol_render_webgl_texturereplay_defaultshader_.Fragment, _ol_webgl_Fragment_);


  /**
   * @const
   * @type {string}
   */
  _ol_render_webgl_texturereplay_defaultshader_.Fragment.DEBUG_SOURCE = 'precision mediump float;\nvarying vec2 v_texCoord;\nvarying float v_opacity;\n\nuniform float u_opacity;\nuniform sampler2D u_image;\n\nvoid main(void) {\n  vec4 texColor = texture2D(u_image, v_texCoord);\n  gl_FragColor.rgb = texColor.rgb;\n  float alpha = texColor.a * v_opacity * u_opacity;\n  if (alpha == 0.0) {\n    discard;\n  }\n  gl_FragColor.a = alpha;\n}\n';


  /**
   * @const
   * @type {string}
   */
  _ol_render_webgl_texturereplay_defaultshader_.Fragment.OPTIMIZED_SOURCE = 'precision mediump float;varying vec2 a;varying float b;uniform float k;uniform sampler2D l;void main(void){vec4 texColor=texture2D(l,a);gl_FragColor.rgb=texColor.rgb;float alpha=texColor.a*b*k;if(alpha==0.0){discard;}gl_FragColor.a=alpha;}';


  /**
   * @const
   * @type {string}
   */
  _ol_render_webgl_texturereplay_defaultshader_.Fragment.SOURCE = _ol_.DEBUG_WEBGL ?
    _ol_render_webgl_texturereplay_defaultshader_.Fragment.DEBUG_SOURCE :
    _ol_render_webgl_texturereplay_defaultshader_.Fragment.OPTIMIZED_SOURCE;


  _ol_render_webgl_texturereplay_defaultshader_.fragment = new _ol_render_webgl_texturereplay_defaultshader_.Fragment();


  /**
   * @constructor
   * @extends {ol.webgl.Vertex}
   * @struct
   */
  _ol_render_webgl_texturereplay_defaultshader_.Vertex = function() {
    _ol_webgl_Vertex_.call(this, _ol_render_webgl_texturereplay_defaultshader_.Vertex.SOURCE);
  };
  _ol_.inherits(_ol_render_webgl_texturereplay_defaultshader_.Vertex, _ol_webgl_Vertex_);


  /**
   * @const
   * @type {string}
   */
  _ol_render_webgl_texturereplay_defaultshader_.Vertex.DEBUG_SOURCE = 'varying vec2 v_texCoord;\nvarying float v_opacity;\n\nattribute vec2 a_position;\nattribute vec2 a_texCoord;\nattribute vec2 a_offsets;\nattribute float a_opacity;\nattribute float a_rotateWithView;\n\nuniform mat4 u_projectionMatrix;\nuniform mat4 u_offsetScaleMatrix;\nuniform mat4 u_offsetRotateMatrix;\n\nvoid main(void) {\n  mat4 offsetMatrix = u_offsetScaleMatrix;\n  if (a_rotateWithView == 1.0) {\n    offsetMatrix = u_offsetScaleMatrix * u_offsetRotateMatrix;\n  }\n  vec4 offsets = offsetMatrix * vec4(a_offsets, 0.0, 0.0);\n  gl_Position = u_projectionMatrix * vec4(a_position, 0.0, 1.0) + offsets;\n  v_texCoord = a_texCoord;\n  v_opacity = a_opacity;\n}\n\n\n';


  /**
   * @const
   * @type {string}
   */
  _ol_render_webgl_texturereplay_defaultshader_.Vertex.OPTIMIZED_SOURCE = 'varying vec2 a;varying float b;attribute vec2 c;attribute vec2 d;attribute vec2 e;attribute float f;attribute float g;uniform mat4 h;uniform mat4 i;uniform mat4 j;void main(void){mat4 offsetMatrix=i;if(g==1.0){offsetMatrix=i*j;}vec4 offsets=offsetMatrix*vec4(e,0.0,0.0);gl_Position=h*vec4(c,0.0,1.0)+offsets;a=d;b=f;}';


  /**
   * @const
   * @type {string}
   */
  _ol_render_webgl_texturereplay_defaultshader_.Vertex.SOURCE = _ol_.DEBUG_WEBGL ?
    _ol_render_webgl_texturereplay_defaultshader_.Vertex.DEBUG_SOURCE :
    _ol_render_webgl_texturereplay_defaultshader_.Vertex.OPTIMIZED_SOURCE;


  _ol_render_webgl_texturereplay_defaultshader_.vertex = new _ol_render_webgl_texturereplay_defaultshader_.Vertex();


  /**
   * @constructor
   * @param {WebGLRenderingContext} gl GL.
   * @param {WebGLProgram} program Program.
   * @struct
   */
  _ol_render_webgl_texturereplay_defaultshader_.Locations = function(gl, program) {

    /**
     * @type {WebGLUniformLocation}
     */
    this.u_image = gl.getUniformLocation(
        program, _ol_.DEBUG_WEBGL ? 'u_image' : 'l');

    /**
     * @type {WebGLUniformLocation}
     */
    this.u_offsetRotateMatrix = gl.getUniformLocation(
        program, _ol_.DEBUG_WEBGL ? 'u_offsetRotateMatrix' : 'j');

    /**
     * @type {WebGLUniformLocation}
     */
    this.u_offsetScaleMatrix = gl.getUniformLocation(
        program, _ol_.DEBUG_WEBGL ? 'u_offsetScaleMatrix' : 'i');

    /**
     * @type {WebGLUniformLocation}
     */
    this.u_opacity = gl.getUniformLocation(
        program, _ol_.DEBUG_WEBGL ? 'u_opacity' : 'k');

    /**
     * @type {WebGLUniformLocation}
     */
    this.u_projectionMatrix = gl.getUniformLocation(
        program, _ol_.DEBUG_WEBGL ? 'u_projectionMatrix' : 'h');

    /**
     * @type {number}
     */
    this.a_offsets = gl.getAttribLocation(
        program, _ol_.DEBUG_WEBGL ? 'a_offsets' : 'e');

    /**
     * @type {number}
     */
    this.a_opacity = gl.getAttribLocation(
        program, _ol_.DEBUG_WEBGL ? 'a_opacity' : 'f');

    /**
     * @type {number}
     */
    this.a_position = gl.getAttribLocation(
        program, _ol_.DEBUG_WEBGL ? 'a_position' : 'c');

    /**
     * @type {number}
     */
    this.a_rotateWithView = gl.getAttribLocation(
        program, _ol_.DEBUG_WEBGL ? 'a_rotateWithView' : 'g');

    /**
     * @type {number}
     */
    this.a_texCoord = gl.getAttribLocation(
        program, _ol_.DEBUG_WEBGL ? 'a_texCoord' : 'd');
  };

}
export default _ol_render_webgl_texturereplay_defaultshader_;
